dcos/nodeutil/util.go

```diff
@@ -0,0 +1,325 @@
+package nodeutil
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/dcos/dcos-go/dcos"
+	"github.com/dcos/dcos-go/exec"
+)
+
+// ErrNodeUtil is an error structure raised by exported functions with meaningful error message.
+type ErrNodeUtil struct {
+	msg string
+}
+
+func (e ErrNodeUtil) Error() string {
+	return e.msg
+}
+
+// NodeInfo defines an interface to interact with DC/OS cluster via go methods.
+type NodeInfo interface {
+	DetectIP() (net.IP, error)
+	Role() (string, error)
+	IsLeader() (bool, error)
+	MesosID() (string, error)
+}
+
+// dcosInfo is implementation of NodeInfo interface.
+type dcosInfo struct {
+	sync.Mutex
+	// save cached data
+	cache bool
+
+	// cached data
+	cachedIP       *net.IP
+	cachedRole     string
+	cachedIsLeader *bool
+	cachedMesosID  string
+
+	// caller parameters
+	client              *http.Client
+	detectIPLocation    string
+	roleMasterFile      string
+	roleAgentFile       string
+	roleAgentPublicFile string
+	mesosStateURL       string
+	dnsRecordLeader     string
+}
+
+// NewNodeInfo returns a new instance of dcosInfo.
+func NewNodeInfo(client *http.Client, options ...Option) (NodeInfo, error) {
+	if client == nil {
+		return nil, ErrNodeUtil{"Client paramter cannot be empty"}
+	}
+
+	// setup default mesos state url.
+	stateURL := url.URL{
```

Comment:
too bad this is buried in the constructor. might be nice to extract this URL as a top level const string (`DefaultStateURL` or something).

Also.. if the `/slaves` endpoint gives you the info you need it's probably a cheaper API call